plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/rule/sync/RefreshBasedSyncMechanism.java

```diff
@@ -0,0 +1,159 @@
+
+    /**
+     * Setting to control the run interval of synchronization service
+     */
+    public static final Setting<Long> RULE_SYNC_REFRESH_INTERVAL_SETTING = Setting.longSetting(
+        RULE_SYNC_REFRESH_INTERVAL_SETTING_NAME,
+        RULE_SYNC_REFRESH_INTERVAL_DEFAULT,
+        MIN_SYNC_REFRESH_INTERVAL,
+        Setting.Property.Dynamic,
+        Setting.Property.NodeScope
+    );
+
+    private final Client client;
+    private final ThreadPool threadPool;
+    private long refreshInterval;
+    private volatile Scheduler.Cancellable scheduledFuture;
+    private final RuleEntityParser parser;
+    private final InMemoryRuleProcessingService ruleProcessingService;
+    private final FeatureType featureType;
+    private static final Logger logger = LogManager.getLogger(RefreshBasedSyncMechanism.class);
+
+    /**
+     * Constructor
+     *
+     * @param client
+     * @param threadPool
+     * @param settings
+     * @param clusterSettings
+     * @param parser
+     * @param ruleProcessingService
+     * @param featureType
+     */
+    public RefreshBasedSyncMechanism(
+        Client client,
+        ThreadPool threadPool,
+        Settings settings,
+        ClusterSettings clusterSettings,
+        RuleEntityParser parser,
+        InMemoryRuleProcessingService ruleProcessingService,
+        FeatureType featureType
+    ) {
+        this.client = client;
+        this.threadPool = threadPool;
+        refreshInterval = RULE_SYNC_REFRESH_INTERVAL_SETTING.get(settings);
+        this.parser = parser;
+        this.ruleProcessingService = ruleProcessingService;
+        this.featureType = featureType;
+        clusterSettings.addSettingsUpdateConsumer(RULE_SYNC_REFRESH_INTERVAL_SETTING, this::setRefreshInterval);
+    }
+
+    void doRun() {
+        clearRulesFromInMemoryService();
+        client.prepareSearch(WorkloadManagementPlugin.INDEX_NAME).execute(new ActionListener<SearchResponse>() {
+            @Override
+            public void onResponse(SearchResponse searchResponse) {
+                Arrays.stream(searchResponse.getHits().getHits())
+                    .forEach(hit -> { ruleProcessingService.add(parser.parse(hit.getSourceAsString())); });
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                logger.info("Failed to refresh rules from in-memory service.");
+            }
+        });
+    }
+
+    private void clearRulesFromInMemoryService() {
+        final AttributeValueStoreFactory attributeValueStoreFactory = ruleProcessingService.getAttributeValueStoreFactory();
+        featureType.getAllowedAttributesRegistry().values().stream().forEach(attribute -> {
+            final AttributeValueStore<String, String> attributeValueStore = attributeValueStoreFactory.getAttributeValueStore(attribute);
+            if (attributeValueStore != null) {
+                attributeValueStore.clear();
+            }
+        });
+    }
```

Comment:
What are the implications of clearing rules for every refresh? Does it impact the tagging of requests during that time?

Reply:
Yes it will, the alternate is to introduce lock based clear which I think might be detrimental since it is a periodic running process